src/odemis/gui/cont/fastem_acq.py

```diff
@@ -581,32 +563,231 @@ def _on_alignment_done(self, future):
+
+    def calibrate(self, evt):
+        """
+        Start or cancel the calibrations for all set regions of calibrations (roc) when the calibration
+        button is triggered. A progressive future for all rocs is created.
+        :param evt: (GenButtonEvent) Button triggered.
+        """
+        # check if cancelled
+        if self._tab_data.main.is_acquiring.value:
+            logging.debug("Calibration was cancelled.")
+            align_fastem._executor.cancel()  # all the rest will be handled by on_alignment_done()
+            # FIXME cancelling does not yet display the correct label in the gui
+            return
+
+        # calibrate
+        self._tab_data.main.is_acquiring.unsubscribe(self._on_is_acquiring)
+        # Briefly unsubscribe as don't need to know about this event (is_acquiring = True):
+        # It  would disable the calibration button, but want to be able to still cancel calibration.
+        self._tab_data.main.is_acquiring.value = True  # make sure the acquire/tab buttons are disabled
+        self._tab_data.main.is_acquiring.subscribe(self._on_is_acquiring)
+
+        self._on_calibration_state()  # update the controls in the panel
+
+        futures = {}
+        for roc_num in sorted(self._tab_data.calibration_regions.value.keys()):
+            # check if calibration region (ROC) on scintillator is set
+            roc = self._tab_data.calibration_regions.value[roc_num]
+            if roc.coordinates.value != stream.UNDEFINED_ROI:
+
+                # calculate the center position of the ROC (half field right/bottom
+                # compared to top/left corner in view)
+                l, t, _, _ = roc.coordinates.value  # (l, t, r, b)
+                field_size = (self._tab_data.main.multibeam.resolution.value[0]
+                              * self._tab_data.main.multibeam.pixelSize.value[0],
+                              self._tab_data.main.multibeam.resolution.value[1]
+                              * self._tab_data.main.multibeam.pixelSize.value[1])
+                roc_center = (l + field_size[0] / 2, t + field_size[1] / 2)
+
+                # start calibration
+                f = align.fastem.align(self._main_data_model.ebeam, self._main_data_model.multibeam,
+                                       self._main_data_model.descanner, self._main_data_model.mppc,
+                                       self._main_data_model.stage, self._main_data_model.ccd,
+                                       self._main_data_model.beamshift, self._main_data_model.det_rotator,
+                                       calibrations=self.calibrations, stage_pos=roc_center)
+                t = align.fastem.estimate_calibration_time(self.calibrations)
+                # also handles cancelling and exceptions
+                f.add_done_callback(partial(self._on_calibration_done, roc_num=roc_num))
+                futures[f] = t
+
+        # if futures:
```

Comment:
what happens if futures is an empty dict?

Reply:
The gui is blocked. I am just wondering if the dict can ever be empty.

Reply:
@pieleric In which case would the dict be empty?